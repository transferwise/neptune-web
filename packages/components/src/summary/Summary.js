import classNames from 'classnames';
import React, { cloneElement } from 'react';
import { useIntl } from 'react-intl';
import Types from 'prop-types';
import {
  CheckCircle as CheckCircleIcon,
  PendingCircle as PendingCircleIcon,
} from '@transferwise/icons';
import requiredIf from 'react-required-if';

import Info from './info';
import { Status } from '../common';
import { deprecated } from '../utilities';

import messages from './Summary.messages';
import './Summary.css';

const BadgeIcons = {
  done: CheckCircleIcon,
  pending: PendingCircleIcon,
};

const statusLabels = {
  [Status.NOT_DONE]: 'statusNotDone',
  [Status.DONE]: 'statusDone',
  [Status.PENDING]: 'statusPending',
};

const expiryDate = new Date('03-01-2021');

const Summary = ({
  action,
  as: Element,
  className,
  content,
  description = content,
  help,
  icon,
  illustration,
  info = help,
  status,
  title,
}) => {
  const intl = useIntl();

  let media = illustration;
  if (icon) {
    const iconSize = icon?.props?.size;
    media = iconSize !== 24 ? cloneElement(icon, { size: 24 }) : icon;
  }
  const Badge = status && BadgeIcons[status];

  return (
    <Element
      className={classNames('np-summary d-flex align-items-start', className)}
      aria-label={status && intl.formatMessage(messages[statusLabels[status]])}
    >
      <div className="np-summary__icon">
        {media}
        {Badge && <Badge size={16} filled className={`np-summary-icon__${status}`} />}
      </div>
      <div className="np-summary__body m-l-2">
        <div className="np-summary__title d-flex">
          <strong>{title}</strong>
          {info && (
            <Info
              aria-label={info['aria-label']}
              className="m-l-1 hidden-xs"
              content={info.content}
              presentation={info.presentation}
              title={info.title}
            />
          )}
        </div>

        {description && <div className="np-summary__description">{description}</div>}
        {action && (
          <a
            href={action.href}
            target={action.target}
            onClick={action.callback}
            className="np-summary__action"
            aria-label={action['aria-label']}
          >
            {action.text}
          </a>
        )}
      </div>
      {info && (
        <Info
          aria-label={info['aria-label']}
          className="m-l-2 hidden-sm hidden-md hidden-lg hidden-xl"
          content={info.content}
          presentation={info.presentation}
          size={Info.Size.LARGE}
          title={info.title}
        />
      )}
    </Element>
  );
};

Summary.Status = {
  NOT_DONE: Status.NOT_DONE,
  DONE: Status.DONE,
  PENDING: Status.PENDING,
};

Summary.propTypes = {
  /** Action displayed at the bottom of the Summary */
  action: Types.shape({
    text: Types.node.isRequired,
    href: Types.string.isRequired,
    'aria-label': Types.string,
    target: Types.string,
    callback: Types.func,
  }),
  /** Decides which html element should wrap the Summary */
  as: Types.elementType,
  /** Extra classes applied to Summary */
  className: Types.string,
  /** @DEPRECATED please use description instead */
  content: deprecated(Types.node, {
    component: 'Summary',
    newProp: 'description',
    expiryDate,
  }),
  /** Summary description */
  // eslint-disable-next-line
  description: Types.node,
  /** @DEPRECATED please use info instead */
  help: deprecated(
    Types.shape({
      content: Types.node.isRequired,
      title: Types.node,
    }),
    {
      component: 'Summary',
      newProp: 'info',
      expiryDate,
    },
  ),
  /** Infos displayed on help Icon click inside Popover or Modal */
  // eslint-disable-next-line
  info: Types.shape({
    'aria-label': Types.string,
    content: Types.node.isRequired,
    presentation: Types.oneOf([Info.Presentation.POPOVER, Info.Presentation.MODAL]),
    title: Types.node,
  }),
  /** @DEPRECATED please use icon instead */
  illustration: deprecated(Types.node, {
    component: 'Summary',
    newProp: 'icon',
    expiryDate,
  }),
  /** Main Summary Icon */
  // eslint-disable-next-line
  icon: requiredIf(Types.node, ({ illustration }) => !illustration),
  /** Decides the badge applied to Icon */
  status: Types.oneOf([Summary.Status.NOT_DONE, Summary.Status.DONE, Summary.Status.PENDING]),
  /** Summary title */
  title: Types.node.isRequired,
};

Summary.defaultProps = {
  action: null,
  as: 'div',
  className: null,
  content: null,
  help: null,
  illustration: null,
  status: null,
};

export default Summary;
