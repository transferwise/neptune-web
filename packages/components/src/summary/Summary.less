@import (reference) '@transferwise/neptune-css/src/less/variables/_typography.less';

.np-summary {
  min-width: 280px;

  &__title {
    margin-top: 2px;
    line-height: calc(22 / 16);

    .np-info__small {
      margin-top: 2px;
    }
  }

  &__icon {
    position: relative;
    color: var(--color-primary);

    .np-summary-icon__done {
      color: var(--color-control-positive);
    }

    .np-summary-icon__pending {
      color: var(--color-control-warning);
    }
    // The second icon is the badge
    :nth-child(2) {
      background-color: var(--color-background-screen);
      border-radius: 50%;
      position: absolute;
      right: -8px;
      bottom: -6px;
    }
  }

  &__description {
    margin-top: var(--size-4);
    font-size: @font-size-small;
    line-height: calc(20 / 14);
  }

  &__action {
    margin-top: var(--size-4);
    font-size: @font-size-small;
    line-height: calc(18 / 14);
  }

  & + .np-summary {
    margin-top: var(--size-24);
  }
}
