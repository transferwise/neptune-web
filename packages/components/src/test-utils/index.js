import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 */
function customRender(ui, { locale = DEFAULT_LOCALE, messages = en, ...renderOptions } = {}) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return <Provider i18n={{ locale, messages }}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render };
