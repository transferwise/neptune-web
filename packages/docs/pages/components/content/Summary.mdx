import { Summary } from '@transferwise/components';
import { Home as HomeIcon } from '@transferwise/icons';

import { LiveEditorBlock, GeneratePropsTable } from '../../../utils';
import code from '../../../liveEditorCode/summary.code';

For displaying a list of requirements, some instructions, or perhaps some general information. Always has an icon and title, can also have a content block and a help icon with some popover text.

<div>
  <LiveEditorBlock code={code} scope={{ Summary, HomeIcon }} />
  <GeneratePropsTable componentName="Summary" />
</div>

export const meta = {
  name: 'Summary',
};
