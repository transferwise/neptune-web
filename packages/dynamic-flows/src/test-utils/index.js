import React from 'react';
import { mount } from 'enzyme';
import { Provider } from '@transferwise/components';
import { act } from 'react-dom/test-utils';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(component, locale = 'en-GB', messages = {}) {
  return mount(<Provider i18n={{ locale, messages }}>{component}</Provider>);
}

export { mountWithProviders as mount, getMockFetchPromise, wait };
