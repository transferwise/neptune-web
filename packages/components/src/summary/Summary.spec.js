import React from 'react';
import { render, waitFor } from '../test-utils';
import Summary from './Summary';

describe('Summary', () => {
  it('renders minimal component', () => {
    const { asFragment } = render(<Summary icon={<strong>icon</strong>} title="Hello world" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders full component', async () => {
    let container;
    await waitFor(() => {
      ({ container } = render(
        <Summary
          action={{
            text: 'text',
            href: 'href',
            'aria-label': 'aria-label',
          }}
          description="description"
          info={{
            title: 'title',
            content: 'description',
            'aria-label': 'aria-label',
          }}
          icon={<strong>icon</strong>}
          status={Summary.Status.DONE}
          title="title"
        />,
      ));

      expect(container).toMatchSnapshot();
    });
  });
});
