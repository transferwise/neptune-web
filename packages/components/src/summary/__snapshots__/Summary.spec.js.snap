// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary renders full component 1`] = `
<div>
  <div
    class="np-summary d-flex align-items-start"
  >
    <div
      class="np-summary__icon"
    >
      <strong
        size="24"
      >
        icon
      </strong>
      <span
        aria-hidden="true"
        class="tw-icon tw-icon-check-circle np-summary-icon__done"
        role="presentation"
      >
        <svg
          fill="currentColor"
          focusable="false"
          height="16"
          width="16"
        >
          <path
            clip-rule="evenodd"
            d="M8 16A8 8 0 108 0a8 8 0 000 16zM4.566 7.434L3.434 8.566 6.5 11.63l6.066-6.065-1.132-1.132L6.5 9.37 4.566 7.434z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </div>
    <div
      class="np-summary__body m-l-2"
    >
      <div
        class="np-summary__title d-flex"
      >
        <strong>
          title
        </strong>
        <span
          class="m-l-1 hidden-xs np-info np-info__small"
        >
          <span
            aria-describedby="id"
            class="np-popover d-inline-block"
          >
            <button
              aria-label="aria-label"
              class="btn-unstyled"
              type="button"
            >
              <span
                aria-hidden="true"
                class="tw-icon tw-icon-help-circle "
                role="presentation"
              >
                <svg
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  width="16"
                >
                  <path
                    d="M9 11a1 1 0 11-2 0 1 1 0 012 0zM10.443 4.8A2.3 2.3 0 007.3 3.96L5.136 5.21l.8 1.385L8.1 5.345a.7.7 0 01.7 1.212l-2.165 1.25.8 1.386L9.6 7.943a2.3 2.3 0 00.842-3.142z"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M8 16A8 8 0 108 0a8 8 0 000 16zm0-1.6A6.4 6.4 0 108 1.6a6.4 6.4 0 000 12.8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </span>
          <div
            class="np-size-swapper d-flex"
            style="visibility: hidden;"
          >
            <div>
              <div
                class="np-panel"
                style="position: absolute; left: 0px; top: 0px;"
              >
                <div
                  class="np-panel__arrow"
                />
                <div
                  aria-hidden="true"
                  class="np-popover__content"
                  id="id"
                  role="tooltip"
                >
                  <span
                    class="sr-only"
                  >
                    Tooltip:
                  </span>
                  <div
                    class="h5"
                  >
                    title
                  </div>
                  description
                </div>
              </div>
            </div>
          </div>
        </span>
      </div>
      <div
        class="np-summary__description"
      >
        description
      </div>
      <a
        aria-label="aria-label"
        class="np-summary__action"
        href="href"
      >
        text
      </a>
    </div>
    <span
      class="m-l-2 hidden-sm hidden-md hidden-lg hidden-xl np-info np-info__large"
    >
      <span
        aria-describedby="id"
        class="np-popover d-inline-block"
      >
        <button
          aria-label="aria-label"
          class="btn-unstyled"
          type="button"
        >
          <span
            aria-hidden="true"
            class="tw-icon tw-icon-help-circle "
            role="presentation"
          >
            <svg
              fill="currentColor"
              focusable="false"
              height="24"
              width="24"
            >
              <path
                d="M13.2 16a1.2 1.2 0 11-2.4 0 1.2 1.2 0 012.4 0zM15.196 8.232a3 3 0 00-4.098-1.098L8.5 8.634l1 1.732 2.598-1.5a1 1 0 111 1.732l-2.598 1.5 1 1.732 2.598-1.5a3 3 0 001.098-4.098z"
              />
              <path
                clip-rule="evenodd"
                d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10zm0-2a8 8 0 100-16 8 8 0 000 16z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
      <div
        class="np-size-swapper d-flex"
        style="visibility: hidden;"
      >
        <div>
          <div
            class="np-panel"
            style="position: absolute; left: 0px; top: 0px;"
          >
            <div
              class="np-panel__arrow"
            />
            <div
              aria-hidden="true"
              class="np-popover__content"
              id="id"
              role="tooltip"
            >
              <span
                class="sr-only"
              >
                Tooltip:
              </span>
              <div
                class="h5"
              >
                title
              </div>
              description
            </div>
          </div>
        </div>
      </div>
    </span>
  </div>
</div>
`;

exports[`Summary renders minimal component 1`] = `
<DocumentFragment>
  <div
    class="np-summary d-flex align-items-start"
  >
    <div
      class="np-summary__icon"
    >
      <strong
        size="24"
      >
        icon
      </strong>
    </div>
    <div
      class="np-summary__body m-l-2"
    >
      <div
        class="np-summary__title d-flex"
      >
        <strong>
          Hello world
        </strong>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
